README.md
# Reloader

More control over hot code push reloading for your mobile apps. A replacement for [`mdg:reload-on-resume`](https://github.com/meteor/mobile-packages/blob/master/packages/mdg:reload-on-resume/README.md) with more options and better UX.

As of Meteor 1.3, if you prevent instant reloading on updates, the newest version of the code will be used on your app's next cold start - no reload necessary. This can be achieved with `Reloader.configure({check: false, refresh: 'start'})`. However, you can also:
- Reload on resume, to update to the newest version of the code when the app is returned from the background: see [`refresh`](#refresh) (the launch screen is put back up during such reloads to hide the white screen you get with `mdg:reload-on-resume`)
- On start or resume, leave the launch screen up and wait to see whether there is an update available: see [`check`](#check), [`checkTimer`](#checktimer), and [`idleCutoff`](#idlecutoff)
- Delay removal of the launch screen, to hide the white screen that appears at the beginning of a reload: see [`launchScreenDelay`](#launchscreendelay)

### Contents

- [Configure](#configure)
  - [check](#check)
  - [checkTimer](#checktimer)
  - [refresh](#refresh)
  - [idleCutoff](#idlecutoff)
  - [launchScreenDelay](#launchscreendelay)
- [Helpers](#helpers)
  - [Reloader.updateAvailable.get()](#reloaderupdateavailableget)
  - [{{updateAvailable}}](#updateavailable)
  - [Reloader.reload()](#reloaderreload)
  - [reloader-update](#reloader-update)
- [Development](#development)
  - [Run tests](#run-tests)
  - [Credits](#credits)

### Installation

```sh
meteor add jamielob:reloader
meteor remove mdg:reload-on-resume
```

If you have any calls to `location.reload()` or `location.replace(location.href)` in your app, replace them with `Reloader.reload()`.

## Configure

The default options are shown below. You can override them anywhere in your `client/` folder.

```js
Reloader.configure({
  check: 'everyStart', // Check for new code every time the app starts 
  checkTimer: 3000,  // Wait 3 seconds to see if new code is available
  refresh: 'startAndResume', // Refresh to already downloaded code on both start and resume
  idleCutoff: 1000 * 60 * 10  // Wait 10 minutes before treating a resume as a start
});
```

These default options will make sure that your app is up to date every time a user starts your app, or comes back to it after 10 minutes of being idle. 

Another popular configuration is:

```js
Reloader.configure({
  check: 'firstStart', // Only make an additonal check the first time the app ever starts
  checkTimer: 5000,  // Wait 5 seconds to see if new code is available on first start
  refresh: 'start' // Only refresh to already downloaded code on a start and not a resume
});
```

This will make sure the first time your app is run it is up to date, will download new versions of code while the app is being used, and then only update when the app is next started.

You can have a different configuration for development, for instance:

```js
if (Meteor.isDevelopment) {
  Reloader.configure({
    check: false, // don't check on startup
    refresh: 'instantly' // refresh as soon as updates are available
  });
} else {
  Reloader.configure({
    // production configuration
  });
}
```

### check

When to make additional checks for new code bundles. The app splash screen is shown during the check. Possible values are:

- `everyStart` (default): Check every time the app starts.  Does not include resuming the app, unless the `idleCutOff` has been reached.
- `firstStart`: Check only the first time the app starts (just after downloading it).
- `false`: Never make additional checks and rely purely on code bundles being downlaoded in the background while the app is being used.

### checkTimer

Default: `3000`

How long to wait (in ms) when making additional checks for new file bundles. In future versions of Meteor we will have an API to instantly check if an update is available or not, but until then we need to simply wait to see if a new code bundle is downloaded. Depending on the size of your app bundle and the phone's connection speed, the default three seconds may not be enough - you can increase it if you find that you have new code immediately after starting the app.

### refresh

When to refresh to the latest code bundle if one finished downloading after the end of the `check` period. The app splash screen is shown during the refresh. Possible values are:

- `startAndResume` (default): Refresh to the latest bundle both when starting and resuming the app.
- `start`: Refresh only when the app is started (not resumed).
- `instantly`: Overrides everything else.  If set, your app will have similar behaviour to the default in Meteor, with code updates being refreshed immeidately. The only improvement/difference is that the app's splash screen is displayed during the refresh.

### idleCutoff

Default: `1000 * 60 * 10 // 10 minutes`

How long (in ms) can an app be idle before we consider it a start and not a resume. Applies only when `check: 'everyStart'`. Set to `0` to never check on resume.

### launchScreenDelay

**Planned option for future version. Currently not configurable.**

Default: `100`

How long to wait (in ms) after reload before hiding the launch screen. The goal is to leave it up until your page has finished rendering, so the user does not see a blank white screen. The duration will vary based on your app's render time and the speed of the device. To be more precise, set `launchScreenDelay` to 0 and release the launch screen yourself when the page has rendered. For example, if the only two pages that might be displayed on reload are `index` and `post`, then you would do:

```javascript
launchScreenHandle = Launchscreen.hold()

Template.index.onRendered(() => {
  launchScreenHandle.release()
});

Template.post.onRendered(() => {
  launchScreenHandle.release()
});
```

Or if you have a layout template, you could put a single `.release()` in that template's `onRendered`.

## Helpers

These helpers can help you to have an "Update Now" button.

### A note about using these helpers
Some people have reported having their app rejected during the Apple review process for having an "Update Now" button or similar as opposed to using the refresh on resume behavior that this package provides by default.  If you really want to have an update button when new code is available, make sure you don't push any new code to the server until after your app has been approved. But it's probably safer/better to simply not have an update button at all!

### How to use them anyway

#### Reloader.updateAvailable.get()

`Reloader.updateAvailable` is a reactive variable that returns true when an update has been downloaded.

```js
Reloader.updateAvailable.get(); // Reactively returns true if an update is ready
```

#### {{updateAvailable}}

This package provides a Blaze template helper that retrieves the value of the reactiveVar easily.

```html
{{#if updateAvailable}}
  	<p>Update available!</p>
{{/if}}
```

#### Reloader.reload()

Call `Reloader.reload()` to refresh the page.

#### reloader-update

This package also provides an easy reload event that you can attach to a button that will briefly show the splash screen and update to the latest version. Simply add the `reloader-update` attribute to a button.

```html
{{#if updateAvailable}}
	<a class="button" reloader-update>Tap here to update!</a>
{{/if}}
```

## Development

### Run tests

```bash
git clone git@github.com:jamielob/reloader.git
cd reloader
meteor test-packages ./ --driver-package practicalmeteor:mocha
open localhost:3000
```

### Credits

[Contributors](https://github.com/jamielob/reloader/graphs/contributors)

And thanks to @martijnwalraven for his help with this packages and superb work on Meteor Cordova! 👏


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

[Change log file format](http://keepachangelog.com/)

## [1.3.0] - 2017-11-13
### Updated

- `cordova-plugin-splashscreen` from `4.0.0` to `4.1.0`

## [1.2.2] - 2016-10-05
### Fixed

- Fix reloading when url contains a hash ([#10](https://github.com/jamielob/reloader/issues/10))

## [1.2.0] - 2016-04-26
### Added
- `launchScreenDelay` option
- tests


reloader-tests.js
// NOTE: I think our Reload._onMigrate messes with test app reloading?
// See "Cannot read property '0' of undefined" in browser console
// after editing the code
//
// In order to get test output to update, you may need to
// click the "refresh" header

import { _ } from 'meteor/underscore';

// http://chaijs.com/api/assert/
import { assert } from 'meteor/practicalmeteor:chai';

// http://sinonjs.org/
import sinon from 'sinon'

describe('refresh', function() {

  beforeEach(function() {
    Reloader.reload = sinon.spy(); // todo save original and repair if needed later
  });

  describe('startAndResume', function() {

    before(function() {
      Reloader.configure({
        refresh: 'startAndResume'
      })
    });

    it('reloads on resume when update is available', function() {
      Reloader.updateAvailable.set(true);

      Reloader._onResume();
      assert.isTrue(Reloader.reload.called);
    });

    it("doesn't reload on resume when update isn't available", function() {
      Reloader.updateAvailable.set(false);

      Reloader._onResume();
      
      assert.isFalse(Reloader.reload.called);
    });

  })

  describe('start', function() {

    before(function() {
      Reloader.configure({
        refresh: 'start'
      })
    });

    it("doesn't reload on resume when update is available", function() {
      Reloader.updateAvailable.set(true);

      Reloader._onResume();
      
      assert.isFalse(Reloader.reload.called);
    });

  });
})

describe('check', function() {
  // should call / not call _checkForUpdate
})



package.js
Package.describe({
  name: 'jamielob:reloader',
  version: '1.3.0',
  summary: 'More control over hot code push reloading',
  git: 'https://github.com/jamielob/reloader/',
  documentation: 'README.md'
});

Cordova.depends({
  'cordova-plugin-splashscreen': '4.1.0'
});

Package.onUse(function(api) {
  api.versionsFrom('1.3.1');

  api.use(['ecmascript',
           'check',
           'underscore',
           'reload',
           'templating',
           'reactive-var',
           'tracker',
           'launch-screen'], 'client');

  api.mainModule('reloader.js', 'web.cordova');
  api.mainModule('browser.js', 'web.browser');
  //  when testing, uncomment this line:
  // api.mainModule('reloader.js', 'client');

  api.export('Reloader', 'client');
});

// No way to make this only happen onTest
Npm.depends({
  sinon: '1.17.3'
});

Package.onTest(function(api) {
  api.use('jamielob:reloader', 'client')

  api.use(['ecmascript',
           'underscore',
           'practicalmeteor:mocha'], 'client');

  api.mainModule('reloader-tests.js', 'client');
});


reloader.js
const launchScreen = LaunchScreen.hold();

Reloader = {

  _options: {},

  configure(options) {
    check(options, {
      check: Match.Optional(Match.OneOf('everyStart', 'firstStart', false)),
      checkTimer: Match.Optional(Match.Integer),
      refresh: Match.Optional(Match.OneOf('startAndResume', 'start', 'instantly')),
      idleCutoff: Match.Optional(Match.Integer),
      launchScreenDelay: Match.Optional(Match.Integer),
    });

    _.extend(this._options, options);
  },

  updateAvailable: new ReactiveVar(false),

  prereload() {
    // Show the splashscreen
    navigator.splashscreen.show();

    // Set the refresh flag
    localStorage.setItem('reloaderWasRefreshed', Date.now());
  },

  reload() {
    this.prereload()

    // We'd like to make the browser reload the page using location.replace()
    // instead of location.reload(), because this avoids validating assets
    // with the server if we still have a valid cached copy. This doesn't work
    // when the location contains a hash however, because that wouldn't reload
    // the page and just scroll to the hash location instead.
    if (window.location.hash || window.location.href.endsWith("#")) {
      window.location.reload();
    } else {
      window.location.replace(window.location.href);
    }
  },


  // Should check if a cold start and (either everyStart is set OR firstStart
  // is set and it's our first start)
  _shouldCheckForUpdateOnStart() {
    const isColdStart = !localStorage.getItem('reloaderWasRefreshed');
    return isColdStart &&
    (
      this._options.check === 'everyStart' ||
      (
        this._options.check === 'firstStart' &&
        !localStorage.getItem('reloaderLastStart')
      )
    );
  },

  // Check if the idleCutoff is set AND we exceeded the idleCutOff limit AND the everyStart check is set
  _shouldCheckForUpdateOnResume() {
    // In case a pause event was missed, assume it didn't make the cutoff
    if (!localStorage.getItem('reloaderLastPause')) {
      return false;
    }

    // Grab the last time we paused
    const lastPause = Number(localStorage.getItem('reloaderLastPause'));

    // Calculate the cutoff timestamp
    const idleCutoffAt = Number( Date.now() - this._options.idleCutoff );

    return (
      this._options.idleCutoff &&
      lastPause < idleCutoffAt &&
      this._options.check === 'everyStart'
    );
  },

  _waitForUpdate(computation) {
    // Check if we have a HCP after the check timer is up
    Meteor.setTimeout(() => {

      // If there is a new version available
      if (this.updateAvailable.get()) {

        this.reload();

      } else {

        // Stop waiting for update
        if (computation) {
          computation.stop()
        }

        launchScreen.release();

      }

    }, this._options.checkTimer );
  },

  _checkForUpdate() {
    if (this.updateAvailable.get()) {

      // Check for an even newer update
      this._waitForUpdate()

    } else {

      // Wait until update is available, or give up on timeout
      Tracker.autorun((c) => {

        if (this.updateAvailable.get()) {
          this.reload();
        }

        this._waitForUpdate(c)

      });

    }
  },

  _onPageLoad() {
    if (this._shouldCheckForUpdateOnStart()) {

      this._checkForUpdate();

    } else {

      Meteor.setTimeout(function() {

        launchScreen.release();

        // Reset the reloaderWasRefreshed flag
        localStorage.removeItem('reloaderWasRefreshed');

      }, this._options.launchScreenDelay); // Short delay helps with white flash

    }
  },

  _onResume() {
    const shouldCheck = this._shouldCheckForUpdateOnResume();

    localStorage.removeItem('reloaderLastPause');

    if (shouldCheck) {

      navigator.splashscreen.show();

      this._checkForUpdate();

      // If we don't need to do an additional check
    } else {

      // Check if there's a new version available already AND we need to refresh on resume
      if ( this.updateAvailable.get() && this._options.refresh === 'startAndResume' ) {

        this.reload();

      }

    }
  },

  // https://github.com/meteor/meteor/blob/devel/packages/reload/reload.js#L104-L122
  _onMigrate(retry) {
    if (this._options.refresh === 'instantly') {

      this.prereload()

      return [true, {}];

    } else {

      // Set the flag
      this.updateAvailable.set(true);

      // Don't refresh yet
      return [false];

    }
  }

};

// Set the defaults
Reloader.configure({
  check: 'everyStart',
  checkTimer: 3000,
  refresh: 'startAndResume',
  idleCutoff: 1000 * 60 * 10, // 10 minutes
  launchScreenDelay: 100
});


Reloader._onPageLoad();

// Set the last start flag
localStorage.setItem('reloaderLastStart', Date.now());

// Watch for the app resuming
document.addEventListener("resume", function () {
  Reloader._onResume();
}, false);


localStorage.removeItem('reloaderLastPause');

// Watch for the device pausing
document.addEventListener("pause", function() {
  // Save to localStorage
  localStorage.setItem('reloaderLastPause', Date.now());
}, false);


// Capture the reload
Reload._onMigrate('jamielob:reloader', function (retry) {
  return Reloader._onMigrate(retry);
});


// Update available template helper
Template.registerHelper("updateAvailable", function() {
  return Reloader.updateAvailable.get();
});

// Update available event
$(document).on('click', '[reloader-update]', function(event) {
  Reloader.reload();
});
